### PR TITLE
Introduce more re-usable helper methods

### PR DESCRIPTION
**Pull request by piontec**
This PR refactors some fixtures to extract reusable helper functions.

### CODE REVIEW

File: pytest_helm_charts/giantswarm_app_platform/app.py [hunk 1]
```diff
@@ -36,65 +30,55 @@ def _app_factory(
...
-        assert app_name != ""
-        assert app_version != ""
-        assert catalog_name != ""
-        assert catalog_url != ""
+        Args:
+             app_name: name of the app in the app catalog
+             app_version: version of the app to use from the app catalog
+             catalog_name: a name of the catalog used for the
+             [AppCatalogCR](pytest_helm_charts.giantswarm_app_platform.custom_resources.AppCatalogCR);
+             new catalog is created only with one with the same name doesn't already exist
```
Comment (rossf7) [inline pytest_helm_charts/giantswarm_app_platform/app.py:40]:
```suggestion
             new catalog is created only when one with the same name doesn't already exist
```

File: pytest_helm_charts/giantswarm_app_platform/app.py [hunk 2]
```diff
@@ -36,65 +30,55 @@ def _app_factory(
...
+             catalog_url: URL of the catalog to install the application from; this is used only if a catalog
+             with the same name doesn't already exists (then a new catalog with the given name and URL is created
+             in the k8s API)
+             namespace: namespace where the App CR will be created
+             deployment_namespace: namespace where the app will be deployed (can be different than `namespace`)
```
Comment (rossf7) [inline pytest_helm_charts/giantswarm_app_platform/app.py:45]: A bit off topic but we should change .spec.namespace in the App CR in a future namespace.

Deployment is a better name we also sometimes call it the target namespace.

File: pytest_helm_charts/giantswarm_app_platform/utils.py [hunk 3]
```diff
@@ -1,22 +1,33 @@
...
 
-def _app_deployed(app: AppCR) -> bool:
+def _app_has_status(app: AppCR, status: str) -> bool:
     complete = (
         "status" in app.obj
         and "release" in app.obj["status"]
         and "appVersion" in app.obj["status"]
         and "status" in app.obj["status"]["release"]
-        and app.obj["status"]["release"]["status"] == "deployed"
+        and app.obj["status"]["release"]["status"].lower() == status
```
Comment (rossf7) [inline pytest_helm_charts/giantswarm_app_platform/utils.py:18]:
```suggestion
        and app.obj["status"]["release"]["status"] == status
```

Do we still need this? With helm 3 the statuses are always lower case.

File: tests/apps/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                    "release": {"status": "deleted"},
+                    "appVersion": "v1",
+                }
+            },
+            None,
+            True,
+        ),
+        # App already doesn't exist
+        (None, pykube.exceptions.ObjectDoesNotExist, True),
+        # Timeout, app exists with unexpected state
```
Comment (rossf7) [inline tests/apps/test_utils.py:62]: I don't get this part since I don't know Python testing. How are timeouts handled in the test?

**Review by rossf7 — COMMENTED** [review]
Nice improvements. From what I understood it looks good :)

Just a few comments and queries.

File: pytest_helm_charts/giantswarm_app_platform/utils.py (hunk 3 above)
Comment (piontec) [reply pytest_helm_charts/giantswarm_app_platform/utils.py:18]: I would keep it; it's not a big performance impact and I remember this being true for Helm2. Maybe some people (outside GS) still use that?

File: tests/apps/test_utils.py (hunk 4 above)
Comment (piontec) [reply tests/apps/test_utils.py:62]: Every tuple inside `(...)` is a sub-test scenario; the sequence of values in the tuple is mapped to parameter values according to the text in the first line above. Then, the test runs N times, each time passing different parameters set as test function arguments.

File: pytest_helm_charts/giantswarm_app_platform/utils.py (hunk 3 above)
Comment (rossf7) [reply pytest_helm_charts/giantswarm_app_platform/utils.py:18]: Yes Helm 2 uses upper case. OK sure that's fine to keep it.

File: pytest_helm_charts/giantswarm_app_platform/utils.py [hunk 5]
```diff
@@ -1,30 +1,192 @@
...
     missing_ok: bool = False,
 ) -> List[AppCR]:
+    """
+    Block until all the apps are running or timeout is reached.
+
+    Args:
+        kube_client: client to use to connect to the k8s cluster
+        app_names: a list of application names to check; all of the applications must be running for this
+        function to succeed
+        app_namespace: namespace where the App CRs of all the apps are stored
```
Comment (rossf7) [inline pytest_helm_charts/giantswarm_app_platform/utils.py:45]:
```suggestion
        app_namespace: namespace where all the App CRs are stored
```

File: pytest_helm_charts/giantswarm_app_platform/utils.py [hunk 6]
```diff
@@ -1,30 +1,192 @@
...
+    app_namespace: str,
+    timeout_sec: int,
+) -> bool:
+    """
+    Block until an App CR has status `deleted` or doesn't exist in k8s API.
+
+    Args:
+        kube_client: client to use to connect to the k8s cluster
+        app_name: an application name to check
+        app_namespace: namespace where the App CRs of all the apps are stored
```
Comment (rossf7) [inline pytest_helm_charts/giantswarm_app_platform/utils.py:78]:
```suggestion
        app_namespace: namespace where all the App CRs are stored
```

File: pytest_helm_charts/giantswarm_app_platform/utils.py [hunk 7]
```diff
@@ -1,30 +1,192 @@
...
+    app_name: str,
+    app_version: str,
+    catalog_name,
+    namespace: str,
+    deployment_namespace: str,
+    config_values: YamlDict = None,
+    namespace_config_annotations: YamlDict = None,
+    namespace_config_labels: YamlDict = None,
+) -> ConfiguredApp:
+    """Creates and deploys a new app using App CR. Optionally creates configuration ConfigMap. Calls are blocking.
```
Comment (rossf7) [inline pytest_helm_charts/giantswarm_app_platform/utils.py:123]:
```suggestion
    """Creates and deploys a new app using App CR. Optionally creates a values ConfigMap. Calls are blocking.
```

In docs we usually call this the values configmap.

File: pytest_helm_charts/utils.py [hunk 8]
```diff
@@ -24,6 +24,33 @@ def wait_for_namespaced_objects_condition(
...
+
+    Args:
+        kube_client: client to use to connect to the k8s cluster
+        obj_type: type of the objects to check; they most be derived from
+        [NamespacedAPIObject](pykube.objects.NamespacedAPIObject)
+        obj_names: a list of object resource names to check; all of the objects must be pass `obj_condition_fun`
+        for this function to succeed
+        objs_namespace: namespace where all the resources are created (single namespace for all resources)
+        timeout_sec: timeout for the call
+        obj_condition_fun: a function that gets one instance of the resource object of type `obj_type`
```
Comment (rossf7) [inline pytest_helm_charts/utils.py:38]:
```suggestion
        obj_condition_func: a function that gets one instance of the resource object of type `obj_type`
```

Comment (rossf7) [reply pytest_helm_charts/utils.py:38]: Is this a typo?

File: tests/apps/test_utils.py (hunk 4 above)
Comment (rossf7) [reply tests/apps/test_utils.py:62]: OK cool thanks for the explanation :)

**Review by rossf7 — APPROVED** [review]
Few final things but LGTM

With the caveat that I don't know Python ;)

File: pytest_helm_charts/utils.py (hunk 8 above)
Comment (piontec) [reply pytest_helm_charts/utils.py:38]: Yes, but consistent across the whole lib! :( Need to fix that

**Pull request merged**